beater/api/firehose/handler.go

```diff
@@ -0,0 +1,160 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package firehose
+
+import (
+	"context"
+	b64 "encoding/base64"
+	"encoding/json"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/beater/auth"
+	"github.com/elastic/apm-server/beater/headers"
+	"github.com/elastic/apm-server/beater/request"
+	"github.com/elastic/apm-server/model"
+	"github.com/elastic/apm-server/publish"
+)
+
+// RequestMetadataFunc is a function type supplied to Handler for extracting
+// metadata from the request.
+type RequestMetadataFunc func(*request.Context) model.APMEvent
+
+type Record struct {
+	Data string `json:"data"`
+}
+
+type FirehoseLog struct {
+	RequestID string   `json:"requestId"`
+	Timestamp int64    `json:"timestamp"`
+	Records   []Record `json:"records"`
+}
+
+// Handler returns a request.Handler for managing firehose requests.
+func Handler(requestMetadataFunc RequestMetadataFunc, processor model.BatchProcessor, authenticator *auth.Authenticator) request.Handler {
+	handle := func(c *request.Context) (*result, error) {
+		accessKey := c.Request.Header.Get("X-Amz-Firehose-Access-Key")
+		if accessKey == "" {
+			return nil, requestError{
+				id:  request.IDResponseErrorsUnauthorized,
+				err: errors.New("Access key is required for Firehose"),
+			}
+		}
+
+		details, authorizer, err := authenticator.Authenticate(context.Background(), headers.APIKey, accessKey)
+		if err != nil {
+			return nil, requestError{
+				id:  request.IDResponseErrorsUnauthorized,
+				err: errors.New("authentication failed"),
+			}
+		}
+
+		c.Authentication = details
+		c.Request = c.Request.WithContext(auth.ContextWithAuthorizer(c.Request.Context(), authorizer))
+		if c.Request.Method != http.MethodPost {
+			return nil, requestError{
+				id:  request.IDResponseErrorsMethodNotAllowed,
+				err: errors.New("only POST requests are supported"),
+			}
+		}
+
+		baseEvent := requestMetadataFunc(c)
+
+		var firehose FirehoseLog
+		err = json.NewDecoder(c.Request.Body).Decode(&firehose)
+		if err != nil {
+			return nil, err
+		}
+
+		var batch model.Batch
+		for _, record := range firehose.Records {
+			recordDec, err := b64.StdEncoding.DecodeString(record.Data)
+			if err != nil {
+				return nil, err
+			}
+
+			splitLines := strings.Split(string(recordDec), "\n")
+			for _, line := range splitLines {
+				if line == "" {
+					break
+				}
+
+				event := baseEvent
+				event.Timestamp = time.Unix(firehose.Timestamp/1000, 0)
+				event.Processor = model.LogProcessor
+				event.Firehose = &model.Firehose{
+					Message: line,
+					ARN:     c.Request.Header.Get("X-Amz-Firehose-Source-Arn"),
+				}
```

Comment:
```suggestion
```

We should not be introducing `model.Firehose`, or `model.APMEvent.Firehose`. The types and fields should (ideally) be a 1:1 mapping with the document structure. Eventually we want to generate the code used for encoding document JSON.

- We already have `APMEvent.Message` -- please use that.
- We should parse the ARN here, and set the APMEvent.Agent and APMEvent.Cloud fields. Actually, this is a good example of "request metadata": ARN is part of the HTTP request: we should do this once for the payload, and initialise a base event which is copied for each log record.